lambda/tmc.ml

```diff
@@ -0,0 +1,854 @@
+           not in tail-call position (after it returns
+           we need to remove the exception handler),
+           so it is not transformed here *)
+        let l1 = traverse ctx l1 in
+        let+ l2 = choice ctx ~tail l2 in
+        Ltrywith (l1, id, l2)
+    | Lstaticcatch (l1, ids, l2) ->
+        (* In [static-catch l1 with ids -> l2],
+           the term [l1] is in fact in tail-position *)
+        let+ l1 = choice ctx ~tail l1
+        and+ l2 = choice ctx ~tail l2 in
+        Lstaticcatch (l1, ids, l2)
+    | Levent (lam, lev) ->
+        let+ lam = choice ctx ~tail lam in
+        Levent (lam, lev)
+    | Lifused (x, lam) ->
+        let+ lam = choice ctx ~tail lam in
+        Lifused (x, lam)
+
+  and choice_apply ctx ~tail apply =
+    let exception No_tmc in
+    try
+      match apply.ap_func with
+      | Lvar f ->
+          let explicit_tailcall_request =
+            match apply.ap_tailcall with
+            | Default_tailcall -> false
+            | Tailcall_expectation true -> true
+            | Tailcall_expectation false ->
+                (* [@tailcall false] disables TMC optimization
+                   on this tailcall *)
+                raise No_tmc
+          in
+          let specialized =
+            try Ident.Map.find f ctx.specialized
+            with Not_found ->
+              if tail then
+                Location.prerr_warning
+                  (Debuginfo.Scoped_location.to_location apply.ap_loc)
+                  Warnings.Tmc_breaks_tailcall;
+              raise No_tmc
+          in
+          {
+            Choice.dps = Dps.make (fun ~tail ~dst ->
+              Lapply { apply with
+                       ap_func = Lvar specialized.dps_id;
+                       ap_args = add_dst_args dst apply.ap_args;
+                       ap_tailcall =
+                         if tail
+                         then Tailcall_expectation true
+                         else Default_tailcall;
+                     });
+            direct = (fun () -> Lapply apply);
+            explicit_tailcall_request;
+            has_tmc_calls = true;
+            benefits_from_dps = true;
+          }
+      | _nontail -> raise No_tmc
+    with No_tmc -> Choice.return (Lapply apply)
+
+  and choice_makeblock ctx ~tail:_ (tag, flag, shape) blockargs loc =
+    let choices = List.map (choice ctx ~tail:false) blockargs in
+    match Choice.find_nonambiguous_tmc_call choices with
+    | Choice.Ambiguous subterms ->
+        let subterms = List.map Choice.direct subterms in
+        raise (Error (Debuginfo.Scoped_location.to_location loc,
+                      Ambiguous_constructor_arguments subterms))
+    | Choice.No_tmc_call args ->
+        Choice.return @@ Lprim (Pmakeblock (tag, flag, shape), args, loc)
+    | Choice.Nonambiguous { Choice.rev_before; choice; after } ->
+        let con = Constr.{
+            tag;
+            flag;
+            shape;
+            before = List.rev rev_before;
+            after;
```

Comment:
There should be a test case that really runs with before and after that are longer than 1.